Dish create page

luciademoja opened the pull request:
PT Story: https://www.pivotaltracker.com/story/show/128658187

Changes proposed in this pull request:
- Adds dish class 
- Sets name and price as required fields
- Includes optional fields
- Shows error when trying to save without filling in name and price

What I have learned working on this feature:
Manage the controller and connect the dots between all the files generated when working with a new model.

Screenshots:
![skarmavbild 2016-08-26 kl 11 49 24](https://cloud.githubusercontent.com/assets/17983955/18001423/3aea31be-6b83-11e6-8b44-7875de908444.png)

Ready for review!

Commit:
Setup of dish model

Commit:
Setup of dish model

Commit:
Added views and form detail,
Added dish to routes
Added 1st Scenario not passing due to no controller setup
Added tables to DB
Added step definition file

Commit:
added create dishes files

luciademoja committed:
small refactoring on dish model changing Dishes class to Dish

luciademoja committed:
adds dishes_controller

luciademoja committed:
adds scenario creates a dish

luciademoja committed:
adds scenario reject when missing required values

tochman commented on spec/models/dish_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Dish, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
A workflow comment: At this point you should have written specs for the Dish model.

tochman replied:
Also, why do we need a Dish model at this stage at all?

luciademoja commented:
This needs to wait, I forgot to work on the last task before doing submitting the PR
Sorry 👎

tochman commented on config/routes.rb:
```diff
@@ -5,4 +5,5 @@
   resources :restaurants, only: [:new, :create]
   root controller: :landing, action: :index
   resources :menus, only: [:new, :create]
+  resources :dishes, only: [:new, :create]
```
This route should have been created when needed, meaning that there should be an acceptance test that tries to access it before you actually create it.

tochman commented on app/models/dishes.rb:
```diff
@@ -0,0 +1,6 @@
+  validates_presence_of :name,
+                        :price,
+
+end
```
At this point you should definitely have specs for Dish

tochman commented on spec/factories/dishes.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :dish do
+    ActiveRecord ""
```
??

tochman commented on app/models/dishes.rb:
```diff
@@ -0,0 +1,6 @@
+class Dishes < ApplicationRecord
```
Dishes?? Should be Dish

tochman commented on app/models/dishes.rb:
```diff
@@ -0,0 +1,6 @@
+class Dishes < ApplicationRecord
+  belongs_to :menu
+  validates_presence_of :name,
+                        :price,
```
The last comma?? Should throw you an error and a spec would have exposed ut ;-)

tochman commented on features/create_dishes.feature:
```diff
@@ -0,0 +1,14 @@
+    And I fill "Incredients" with "Mince"
+    And I fill "Calories" with "429"
+    And I click "Create"
+    Then I should be on the "home" page
```
Good thing. Next time START with that

tochman commented on app/views/dishes/new.html.haml:
```diff
@@ -0,0 +1,19 @@
+    = f.text_area :allergy_info
+    .field
+      = f.label :price
+      = f.text_field :price
```
Redundant? Two fields for :price?

tochman commented on app/controllers/dishes_controller.rb:
```diff
@@ -0,0 +1,23 @@
+  end
+
+  # def create
+  #   @menu = Dishes.new(dish_params)
```
Cut and paste is not cool ;-)

tochman commented on app/controllers/dishes_controller.rb:
```diff
@@ -3,21 +3,21 @@ def new
+    if @dish.save
+      redirect_to root_path
+    else
+      flash[:error] = 'A name is required'
```
I would use the validation messages from the model. You can access them by: 

```
@dish.errors.full_messages
```

tochman commented:
@lollypop27 && @luciademoja Good job on this feature. My comments are mostly about the workflow and not about the content. I just want you to follow the Acceptance-Unit Test Cycle more strict. ;-)

diraulo commented:
@luciademoja we have a merge conflict here

MikaelFeher committed:
Fixes merging conflicts

MikaelFeher committed:
Adds unittests for dish model

tochman committed:
refactores dish create route and new & create methods to require menu
adds has_many realtionship between menu & dishes
updates specs
updates form with @menu object

tochman commented:
@luciademoja Me and @MikaelFeher worked a bit on this code and created a PR toward your fork of this repo. Have a look at that and merge it in if you are confident with those proposed changes.

MikaelFeher committed:
Dish create (#1)

* Restaurant edit details page.  (#24)

* Adds an edit action, view and routes to restaurants
refactoring of some step definitions
* Adds function to update restaurant details
* minor refactoring
* Adds fallback in case of failed update
refactored feature test and step definitions
* Adds partial to DRY up the new and edit views
Adds minor refactoring to feature test
* Adds more detailed error messages
Renamed form partial
Minor refactoring
* Refactored error message code to fix display issue
updated feature test to display correct error message
* Refactors method name for flash messages
* Adds refactoring to step definitions

* removes obsolete jquery file (#25)

* Establish Restaurant gelocation by street adress (#26)

* adds lat, lng & city to restaurant
adds geocoder gem
updates specs for restaurant
creates a stub file for geocoder addresses to be used in testing (1 address only)
updates rails_helper and env.rb

* updates view and features with city attribute

* adds lat & lng to restaurant after validation

* changes 'after_validation' to 'after_save' for geocode call

* Adds unittests for dish model

* refactores dish create route and new & create methods to require menu
adds has_many realtionship between menu & dishes
updates specs
updates form with @menu object

* Adds error messages
Minor refactoring

tochman committed:
fixes merging conflicts

luciademoja committed:
fixes merging conflicts with upstream develop

luciademoja committed:
updated with latest PR from Mikael

luciademoja committed:
fixes latest conflicts

luciademoja committed:
cleanes the app from the automatically generated files when scaffolding

diraulo commented on features/step_definitions/create_menu_page.rb:
```diff
@@ -16,8 +16,17 @@
       path = new_user_registration_path
     when 'restaurants' then
       path = new_restaurant_path
+    when 'create dish' then
```
Is this really necessary? You have a step that visits the create dish page and the path is set there.

tochman replied:
That step def is a bit messy. Leave this for the refactoring of step defs chore?

MikaelFeher replied:
Yes, leave it, there is much work to be done in the definitions.

diraulo replied:
Okay then if you have a chore to refactor the step defs.. In general the less code we pull in the first place the less we have to refactor later on. But I will let this one fly for now.

tochman commented:
Lucia is still working on a task. Not really ready to merge yet

luciademoja committed:
tries to update the schema with the category param for dish

tochman committed:
adds categoriy to dish
adds validation (need a spec)
adds a constant to Dish with predefined categories and updates the template element

tochman committed:
adds spec for custom category

luciademoja committed:
adds category param to dish

tochman commented on app/models/dish.rb:
```diff
@@ -1,11 +1,11 @@
                         :category
 
-  validates :category, inclusion: ['Starter', 'Mail Course', 'Dessert']
+  validates :category, inclusion: ['Starter', 'Main Course', 'Dessert']
```
```
 validates :category, inclusion: CATEGORIES
```

luciademoja committed:
small refactoring on dish.rb

diraulo commented:
@luciademoja @tochman this one still says WIP?
